core/src/main/java/org/elasticsearch/action/support/ListenerExecutionContext.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ListenerExecutionContext<V> implements ActionListener<V> {
+
+    private static Object NULL_VALUE = new Object();
+
+    private final ConcurrentLinkedQueue<ActionListener<V>> listeners = new ConcurrentLinkedQueue<>();
+    private final AtomicReference<Object> result = new AtomicReference<>(null);
+
+    @Override
+    public void onResponse(V value) {
+        if (result.compareAndSet(null, value != null ? value : NULL_VALUE)) {
+            ActionListener<V> listener;
+            while ((listener = listeners.poll()) != null) {
+                listener.onResponse(value);
+            }
+        }
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        if (e == null) {
+            throw new IllegalArgumentException("Exception cannot be null");
+        }
+        if (result.compareAndSet(null, e)) {
+            ActionListener<V> listener;
+            while ((listener = listeners.poll()) != null) {
+                listener.onFailure(e);
+            }
+        }
+    }
+
+    public boolean isDone() {
+        return result.get() != null;
+    }
+
+    public void addListener(ActionListener<V> listener) {
+        internalAddListener(listener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void internalAddListener(ActionListener<V> listener) {
+        listeners.offer(listener);
```

Comment:
we have a race here. We might execute this listener twice. I think we should keep it simple and just use `synchronize` and make a fully copy of the list every time we add a listener. I don't think this is perf critical. Let's make the state only mutable under a mutex otherwise it's too complicated. That way you can also just use `ActionListener#onResponse` and `ActionListener#onFailure`